### ISSUE TITLE
trust-store broken on devel (Nexus 4)

### ISSUE THREAD

**Issue by mardy** [ubports/ubuntu-touch#1719]
- Device: mako
- Channel: devel
- Build: 2021-04-20

### Steps to reproduce
Open an application which requires access to camera, microphone or location (and that has never asked for access below) in order to trigger the trust-store prompt.

### Expected behavior
The prompt appears.

### Actual behavior
The screen gets black (except for the top Lomiri panel), and either the device reboots after a few seconds, or the trust-prompt process uses 100% of the CPU.

### Logfiles and additional information
These messages appear in the syslog:

```
Apr 21 14:03:33 ubuntu-phablet kernel: [   70.352571] kgsl kgsl-3d0: |kgsl_iommu_fault_handler| GPU PAGE FAULT: addr = C0010000 pid = 3742
Apr 21 14:03:33 ubuntu-phablet kernel: [   70.352846] kgsl kgsl-3d0: |kgsl_iommu_fault_handler| context = 0 FSR = 4001000A FSYNR0 = F000008 FSYNR1 = 443542(read fault)
Apr 21 14:03:33 ubuntu-phablet kernel: [   70.353059] kgsl kgsl-3d0: ---- nearby memory ----
Apr 21 14:03:33 ubuntu-phablet kernel: [   70.353212] kgsl kgsl-3d0: [C0004000 - C0005000] (+guard) (pid = 3742) (egl_surface)
Apr 21 14:03:33 ubuntu-phablet kernel: [   70.353425] kgsl kgsl-3d0:  <- fault @ C0010000
Apr 21 14:03:33 ubuntu-phablet kernel: [   70.353639] kgsl kgsl-3d0: *EMPTY*
Apr 21 14:03:33 ubuntu-phablet kernel: [   70.445780] kgsl kgsl-3d0: trust-prompt[3742]: gpu fault ctx 9 ts 7 status F5C7C003 rb 0350/049b ib1 ae430ef8/06a9 ib2 ad374000/0000
Apr 21 14:03:33 ubuntu-phablet kernel: [   70.559774] kgsl kgsl-3d0: |adreno_snapshot| GPU snapshot froze 140Kb of GPU buffers
Apr 21 14:03:33 ubuntu-phablet kernel: [   70.560842] kgsl kgsl-3d0: |kgsl_device_snapshot| snapshot created at pa af900000 size 142712
Apr 21 14:03:33 ubuntu-phablet kernel: [   70.561452] kgsl kgsl-3d0: |kgsl_iommu_clk_disable_event| IOMMU disable clock event being cancelled, iommu_last_cmd_ts: ec1, retired ts: ebc
Apr 21 14:03:33 ubuntu-phablet kernel: [   70.567526] kgsl kgsl-3d0: |kgsl_snapshot_save_frozen_objs| Allocated memory for snapshot objects at address f0376000, size 7778
Apr 21 14:03:33 ubuntu-phablet kernel: [   70.612757] kgsl kgsl-3d0: trust-prompt[3742]: gpu failed ctx 9 ts 7
Apr 21 14:03:53 ubuntu-phablet kernel: [   90.602685] kgsl kgsl-3d0: |adreno_reset| Device soft reset failed
Apr 21 14:03:53 ubuntu-phablet kernel: [   90.602960] kgsl kgsl-3d0: |kgsl_iommu_clk_disable_event| IOMMU disable clock event being cancelled, iommu_last_cmd_ts: eca, retired ts: ec4
Apr 21 14:03:53 ubuntu-phablet kernel: [   90.652739] kgsl kgsl-3d0: unity8[3306]: gpu fault ctx 6 ts 845 status C000C001 rb 0040/0091 ib1 98017e80/017d ib2 00000000/0000
```

Note: this does not happen on the stable channel.

Tagging @UniversalSuperBox as he reported having seem the same issue. :-)

**Comment by UniversalSuperBox** [comment]
I've found that this issue does not occur on the Nexus 5, Nexus 7, or Xperia X. It's *just* the Nexus 4 so far.

**Comment by UniversalSuperBox** [comment]
Through careful testing (read: pushing the button over and over again), I was able to determine that the Nexus 7 2013 (WiFi) suffers the same issue: Opening a trust-prompt sometimes causes the entire device to lock up.

Since the Nexus 7 is easier to test with, I continued on it.

After trying unsuccessfully to debug the issue with gdb (instead of locking up, the trust-store application simply goes back to `poll()` as would be expected from a working Qt application), strace (no good pattern found from the output), and apitrace (causes any applications it links to to segfault, thanks libhybris), I tried another route: throw something at the wall and see if it sticks.

Although at this point, we had learned something valuable: if the system doesn't reboot due to the GPU lockup, it continues functioning relatively normally. Applications can still take input, their output just doesn't get rendered.

Alfred suggested I look at the kgsl driver in the kernel. So I took a look at [Google's newest upstream source for flo in the drivers/gpu/msm folder](https://android.googlesource.com/kernel/msm/+log/refs/heads/android-msm-flo-3.4-marshmallow-mr2/drivers/gpu/msm) and compared it to [the current head of our flo branch](https://github.com/ubports/ubuntu_kernel_xenial/commits/e9b6a474701afe0fa7c2b660b7c27557e33ce4b3). Turns out we were quite a few commits behind. I pulled them in successfully with a little magic [1] and rebuilt the kernel. To my surprise, it booted with EGL applications running. To no one's surprise, the trust-store continued locking up. However, the logs got a little longer[2]. It turned out that the GPU was now skipping or failing the exact GL calls that caused the fault rather than locking up completely! Only the application that tried to use trust-store froze while the system continued running. This was progress.

Continuing on, I wondered if this might be related to https://github.com/ubports/ubuntu-touch/issues/1434. It turns out, it is. The GPU reports the same errors when failing to open an online-accounts prompt as it reports for trust-store. And, indeed, after the patches are installed, a failed opening of the online-accounts prompt doesn't cause the system to lock up and reboot, but rather the settings application locks up and a corrupted approximation of the login dialog appears.

![corrupted online-accounts prompt placed side-by-side with a correct online-accounts prompt](https://user-images.githubusercontent.com/21966173/116583253-c6b53280-a8db-11eb-8b0d-51f955507c3a.png)

It's worth noting three things about this corrupted approximation (which was scaled down for size constraints in the comment): 

1. Its dimensions are always 864x1080, even when the prompt is opened with the device rotated to Landscape instead of Portrait (Portrait is shown here).
2. It always accompanies a GPU page fault message with the fault address offset by `0xC40` or `0xC00` from the end of a surface owned by the process that caused the fault [3]
3. The corruption is always accompanied by the page fault message. However, the page fault message is not always accompanied by corruption. It is possible to get a freeze and a page fault message followed by a few `gpu replay` messages and a successfully-rendered dialog. It is also possible to get a freeze and a page fault followed by the dialog rendering normally for one frame before corrupting.

I don't know the correct way forward yet. However, observation (2) lead me to find that all trust-store prompts cause a page fault offset by `0x1800` from the end of a surface owned by the faulting process.

We do know one thing though: this is almost certainly related to the MirPromptSession. That's not a lot to go off of, but it's something.

I'll continue my attempts to locate the real issue until the end of the day. If I'm unable to make more progress, I'll spend tomorrow rebuilding the mako and flo kernels to include the kgsl changes mentioned above and ship those. That should downgrade this issue from `critical` to `bug`, since the system is still usable after the error (you can restart the app and try again until the prompt works). Not perfect, but better than a full system lockup.

[1] `git cherry-pick $(git log --format=oneline --no-merges --reverse HEAD..google/android-msm-flo-3.4-marshmallow-mr2  drivers/gpu/msm/ | cut -d " " -f 1)`

[2]
```
[  105.856567] kgsl kgsl-3d0: |kgsl_iommu_fault_handler| GPU PAGE FAULT: addr = C0004800 pid = 4507
[  105.856719] kgsl kgsl-3d0: |kgsl_iommu_fault_handler| context = 0 FSR = 4001000A FSYNR0 = D000008 FSYNR1 = 443542(write fault)
[  105.856903] kgsl kgsl-3d0: ---- nearby memory ----
[  105.856994] kgsl kgsl-3d0: [C0002000 - C0003000] (+guard) (pid = 4507) (egl_surface)
[  105.857147] kgsl kgsl-3d0:  <- fault @ C0004800
[  105.857238] kgsl kgsl-3d0: *EMPTY*
[  106.022125] kgsl kgsl-3d0: trust-prompt[4507]: gpu fault ctx 9 ts 5 status F5C7C003 rb 0890/090e ib1 ae3879c4/051e ib2 ad2cb000/0000
[  106.135620] kgsl kgsl-3d0: trust-prompt[4507]: gpu skipib ctx 9 ts 5 policy 4
[  106.432189] kgsl kgsl-3d0: trust-prompt[4507]: gpu fault ctx 9 ts 6 status F5C7C003 rb 0070/03e6 ib1 ae389bf8/0348 ib2 ad2cb000/0000
[  106.806335] kgsl kgsl-3d0: trust-prompt[4507]: gpu replay ctx 9 ts 6 policy 6
[  106.951873] kgsl kgsl-3d0: trust-prompt[4507]: gpu fault ctx 9 ts 7 status F5C7C003 rb 0070/04d6 ib1 ae38b960/0522 ib2 ad2cb000/0000
[  106.952117] kgsl kgsl-3d0: trust-prompt[4507]: gpu fault threshold exceeded 3 faults in 3000 msecs
[  106.952362] kgsl kgsl-3d0: trust-prompt[4507]: gpu skipped ctx 9 ts 7
```

```
[ 1035.354248] kgsl kgsl-3d0: |kgsl_iommu_fault_handler| GPU PAGE FAULT: addr = C0796C40 pid = 4957
[ 1035.354339] kgsl kgsl-3d0: |kgsl_iommu_fault_handler| context = 0 FSR = 40010008 FSYNR0 = D00000B FSYNR1 = 443542(write fault)
[ 1035.354492] kgsl kgsl-3d0: ---- nearby memory ----
[ 1035.354675] kgsl kgsl-3d0: [C0400000 - C0796000] (+guard) (pid = 4957) (egl_surface)
[ 1035.354766] kgsl kgsl-3d0:  <- fault @ C0796C40
[ 1035.354919] kgsl kgsl-3d0: *EMPTY*
```

**Comment by UniversalSuperBox** [comment]
It can now be considered "later than the end of the day" and I'm pretty sure this cannot be fixed in a reasonable amount of time proportional to the number of people serviced by the Nexus 4 and 7.

The error we're seeing is, actually, shockingly well-documented across the internet: https://developer.qualcomm.com/forum/qdevnet-forums/mobile-technologies/mobile-gaming-graphics-optimization-adreno/27030 and https://bugs.dolphin-emu.org/issues/10038... it was even experienced by Canonical: https://bugs.launchpad.net/mir/+bug/1566747/

I believe we cannot work around an overload of GL commands as guessed by the people in the Dolphin bug report since we are using double buffering, so `glFinish()` or `glFlush()` have little more effect than just swapping buffers does anyway.

Qualcomm developers seem to think they fixed the issue in a later Adreno release which may or may not be available for the Nexus 4 and 7. If it is fixed, it's likely fixed in one of the drivers for Android 5.1 or 6. These drivers are all available for download from Google, but unfortunately our Android 4.4.2 rootfs can't run an Adreno driver newer than 4.4.4 (and no, the 4.4.4 drivers don't fix the issue).

"If it's fixed" is another important assumption here. The Dolphin bug report is from 2017 and reports using some Wileyfox device (it's impossible to find out which one since the link is offline). Android 7 was released in late 2016... While it's possible that Wileyfox kept shipping Adreno from 2013 in 2017, I find it rather unlikely. So this might be a bug that was never fixed at all!

But everything I just wrote assumes that the Adreno driver is broken and not our code. Unfortunately, determining if this is the case is rather difficult... Qualcomm didn't think to include the KHR_debug extension in this EGL implementation and Qt has no other easy way to log the input and output of every EGL and GL call. Additionally, `apitrace` causes any application it links to to segfault. The best I can do is determine approximately which frame causes the issue, which is almost entirely useless given GL's command queue:

I added some logging to QtUbuntu with [this patch](https://gist.github.com/UniversalSuperBox/0657831f311e2e39c6beed02b01d1797). It shows when a `swapBuffers()` call starts, when it ends, and whether any errors occurred before or after the swap. You'll also need https://github.com/ubports/trust-store/pull/16 if you want to debug trust-store.

A normal trust-store run is pretty boring:

`~/.cache/upstart/pulseaudio-trust-stored.log`
```
BEFORE!
AFTER! 9
BEFORE!
AFTER! 12
BEFORE!
AFTER! 13
BEFORE!
AFTER! 3
```

But sometimes we get a `GL_INVALID_ENUM` (0x0500, 1280) error and the application has a tiny pause. Nothing is logged in dmesg, and a very unhelpful `W/Adreno-ES20( 5498): <core_glTexParameteriv:566>: GL_INVALID_ENUM` is logged by Adreno.

`~/.cache/upstart/pulseaudio-trust-stored.log`
```
BEFORE!
AFTER! 9
BEFORE!
AFTER! 12
BEFORE!
GL ERROR BEFORE! 1280
AFTER! 18
BEFORE!
AFTER! 3
```

And other times, we get the `GL_INVALID_ENUM`, the next buffer swap times out and we get the errors in dmesg outlined in my last comment. When this happens, the egl context receives `EGL_CONTEXT_LOST` (0x300E) and after that `GL_INVALID_OPERATION` (0x0502, 1282) on every further GL call. Makes sense since our entire GL context is lost.

```
BEFORE!
AFTER! 22
BEFORE!
GL ERROR BEFORE! 1280
AFTER! 1080
BEFORE!
QEGLPlatformContext: eglSwapBuffers failed: 300e
AFTER! 0
GL ERROR BEFORE! 1282
GL ERROR BEFORE! 1282
GL ERROR BEFORE! 1282
```

So, without having a good idea of how to track down that `GL_INVALID_ENUM` (which might not even be the real issue in the end), we're kinda stuck. I'll rebuild the kernels tomorrow and we'll ship OTA-17 with this bug, just in a reduced form.

**Comment by UniversalSuperBox** [comment]
Fixed on Nexus 4 with https://gitlab.com/ubports/community-ports/mako/-/commit/fbb61841269ce2d29a934dc419c2efda58e7fd98. Fixed on Nexus 7 with https://gitlab.com/ubports/community-ports/flo/-/commit/6a226d03f5d370b408326662dd722980199c6532. Both commit hashes referenced are in https://github.com/ubports/ubuntu_kernel_xenial.

The issue still persists slightly, but is no longer critical and does not meet the criteria in this issue report. I have filed the issue again as https://github.com/ubports/ubuntu-touch/issues/1722.
